src/libteam/0005-libteam-Add-warm_reboot-mode.patch

```diff
@@ -191,7 +192,22 @@ index 81324de..519f5e2 100644
  	struct {
  		uint32_t speed;
  		uint8_t	duplex;
-@@ -994,6 +996,13 @@ static int lacp_port_set_state(struct lacp_port *lacp_port,
+@@ -496,9 +498,13 @@ static int lacp_update_carrier(struct lacp *lacp)
+ 		err = teamd_port_enabled(lacp->ctx, tdport, &state);
+ 		if (err)
+ 			return err;
+-		if (state && ++ports_enabled >= lacp->cfg.min_ports)
++		if (state && ++ports_enabled >= lacp->cfg.min_ports) {
++			lacp->ctx->warm_start_started = true;
+ 			return lacp_set_carrier(lacp, true);
++		}
+ 	}
++	if (lacp->ctx->warm_start && !lacp->ctx->warm_start_started)
```

Comment:
Thanks for the change,  even if teamd was started with warmboot mode,  we should still set the carrier according to the port enabled state (e,g, number of enabled port less than the min_ports , we should put it down. etc.) during the run time to keep the state right?  Do you think lacp_carrier_init should be able to restore to whatever state before warmboot and we don't need do anything special here?

Reply:
When we start teamd in WR mode we don't have any information about current desired state of the LAG interface.
If the box wasn't restarted, we could read the current state of teamd, and consider that no interfaces wasn't put down during transient period. Then teamd will read saved LACP PDUs, and restore previous state of itself. If some state changed since stop of previous instance of teamd, the new instance of teamd will catch it up eventually (up to 90 seconds in the worst case)
If the box was restarted, we must create LAG interface and change it state to the saved state ASAP. This patch doesn't have this behavior. Current patch doesn't change state the LAG interface

Reply:
Because, if the previous state was down, we don't put the LAG interface up, only because we're in WR mode.

But if the previous state was up, we put the interface up as soon as we get into lacp_update_carrier function, which is fast. And after the interface is up after we have enough links in the group we put teamd behavior to the normal mode.

What is bad in my patch. If the previous state of LAG was down, and after start we have LAG in down state, it'd be up as soon as we add one interface to a group, so no min_links check. I'm thinking currently how I can fix that.